titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/PodToTaskMapper.java

```diff
@@ -0,0 +1,327 @@
+            // Pod in running state that did not complete must have been terminated as well.
+            if (taskState == KillInitiated && task.getStatus().getReasonCode().equals(REASON_STUCK_IN_STATE)) {
+                reasonCode = REASON_TRANSIENT_SYSTEM_ERROR;
+            } else {
+                reasonCode = REASON_TASK_KILLED;
+            }
+        } else if (podWrapper.getPodPhase() == PodPhase.SUCCEEDED) {
+            if (taskState == KillInitiated && task.getStatus().getReasonCode().equals(REASON_STUCK_IN_STATE)) {
+                reasonCode = REASON_TRANSIENT_SYSTEM_ERROR;
+            } else {
+                if (podWrapper.hasDeletionTimestamp() || taskState == KillInitiated) {
+                    reasonCode = REASON_TASK_KILLED;
+                } else {
+                    reasonCode = REASON_NORMAL;
+                }
+            }
+        } else if (podWrapper.getPodPhase() == PodPhase.FAILED) {
+            reasonCode = REASON_FAILED;
+        } else {
+            titusRuntime.getCodeInvariants().inconsistent("Pod: %s has unknown phase mapping: %s", podWrapper.getName(), podWrapper.getPodPhase());
+            reasonCode = REASON_UNKNOWN;
+        }
+
+        return Either.ofValue(TaskStatus.newBuilder()
+                .withState(Finished)
+                .withReasonCode(effectiveFinalReasonCode(reasonCode))
+                .withReasonMessage("Container was terminated without going through the Titus API")
+                .withTimestamp(now)
+                .build()
+        );
+    }
+
+    /**
+     * Handle pod 'Pending' phase. In this phase we have to distinguish cases:
+     * <ul>
+     *     <li>pod is waiting in the queue</li>
+     *     <li>pod is assigned to a node but not running yet</li>
+     * </ul>
+     */
+    private Either<TaskStatus, String> handlePodPending() {
+        if (!podWrapper.isScheduled()) {
+            return handlePodPendingQueued();
+        }
+        return handlePodPendingInitializingInContainerWaitingState(podWrapper);
+    }
+
+    /**
+     * Handle pod is waiting in the queue (is not scheduled yet). This event does not cause any task state
+     * transition. We only verify consistency.
+     */
+    private Either<TaskStatus, String> handlePodPendingQueued() {
+        if (task.getStatus().getState() != TaskState.Accepted) {
+            return unexpected("expected queued task in the Accepted state");
+        }
+        return irrelevant("pod is waiting in the queue to be scheduled");
+    }
+
+    /**
+     * Handle pod is assigned to a node but not running yet. The following pre-conditions must exist:
+     * <ul>
+     *     <li>pod condition 'PodScheduled' is set</li>
+     *     <li>nodeName is not empty</li>
+     * </ul>
+     * Based on container state and reason message the state is classified as 'Launched' or 'StartInitiated':
+     * <ul>
+     *     <li>if reason != 'TASK_STARTING' set 'Launched'</li>
```

Comment:
I think we can use `V1ContainerStateRunning.getStartedAt` here in the future as these special reasons are going away.